cmd/setup-node/commands/root.go

```diff
@@ -60,52 +70,62 @@ var rootCmd = &cobra.Command{
 				log.Fatalf("Failed to open config: %v", err)
 			}
 		} else {
-			logger.Info("Reading config from STDIN")
+			log.Info("Reading config from STDIN")
 			rdr = bufio.NewReader(os.Stdin)
 		}
 
 		conf := &setup.Config{}
 
 		raw, err := ioutil.ReadAll(rdr)
 		if err != nil {
-			logger.Fatalf("Failed to read config: %v", err)
+			log.Fatalf("Failed to read config: %v", err)
 		}
 
 		if err := json.Unmarshal(raw, &conf); err != nil {
-			logger.WithField("raw", string(raw)).Fatalf("Failed to decode config: %s", err)
+			log.WithField("raw", string(raw)).Fatalf("Failed to decode config: %s", err)
 		}
 
-		logger.Infof("Config: %#v", conf)
+		log.Infof("Config: %#v", conf)
 
-		sn, err := setup.NewNode(conf, metrics.NewPrometheus("setupnode"))
+		sn, err := setup.NewNode(conf)
 		if err != nil {
-			logger.Fatal("Failed to create a setup node: ", err)
+			log.Fatal("Failed to create a setup node: ", err)
 		}
 
-		go func() {
-			http.Handle("/metrics", promhttp.Handler())
-			if err := http.ListenAndServe(metricsAddr, nil); err != nil {
-				logger.Println("Failed to start metrics API:", err)
-			}
-		}()
+		m := prepareMetrics(log)
 
-		ctx, cancel := cmdutil.SignalContext(context.Background(), logger)
+		ctx, cancel := cmdutil.SignalContext(context.Background(), log)
 		defer cancel()
 
-		logger.Fatal(sn.Serve(ctx))
+		log.Fatal(sn.Serve(ctx, m))
 	},
 }
 
-func init() {
-	rootCmd.Flags().StringVarP(&metricsAddr, "metrics", "m", ":2121", "address to bind metrics API to")
-	rootCmd.Flags().StringVar(&syslogAddr, "syslog", "", "syslog server address. E.g. localhost:514")
-	rootCmd.Flags().StringVar(&tag, "tag", "setup-node", "logging tag")
-	rootCmd.Flags().BoolVarP(&cfgFromStdin, "stdin", "i", false, "read config from STDIN")
+func prepareMetrics(log logrus.FieldLogger) setupmetrics.Metrics {
+	if metricsAddr == "" {
+		return setupmetrics.NewEmpty()
+	}
+
+	m := setupmetrics.New(tag)
+	mux := http.NewServeMux()
+
+	// TODO: The following should be replaced by promutil.AddMetricsHandle
+	reg := prometheus.NewPedanticRegistry()
+	reg.MustRegister(prometheus.NewProcessCollector(prometheus.ProcessCollectorOpts{}))
+	reg.MustRegister(prometheus.NewGoCollector())
+	reg.MustRegister(m.Collectors()...)
+	h := promhttp.InstrumentMetricHandler(reg, promhttp.HandlerFor(reg, promhttp.HandlerOpts{}))
+	mux.Handle("/metrics", h)
+
+	log.WithField("addr", metricsAddr).Info("Serving metrics...")
+	go func() { log.Fatal(http.ListenAndServe(metricsAddr, mux)) }()
+
+	return m
 }
 
 // Execute executes root CLI command.
 func Execute() {
 	if err := rootCmd.Execute(); err != nil {
-		log.Fatal(err)
+		panic(err)
```

Comment:
Why do you think this change is necessary?

Reply:
`log.Fatal` logs the error and then panics. However, when a `cobra.Command` fails, it already logs the error. So we only need to panic here.